File: revc/revc.py
#!/usr/bin/env python

"""This script is given "A DNA string s of length at most 1000 bp" and
returns "The reverse complement sc of s".
"""

import sys
import argparse

# Ensure maketrans works in Python 2 and 3
if sys.version_info[0] >= 3:
    maketrans = str.maketrans
else:
    import string
    maketrans = string.maketrans


def main(args):
    """Complementing a Strand of DNA"""

    s = args.dataset.read()

    print(s.rstrip().translate(maketrans('ACGT', 'TGCA'))[::-1])

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='Complementing a Strand of DNA')
    parser.add_argument(
        'dataset', metavar='FILE', type=argparse.FileType('r'),
        help='A DNA string s of length at most 1000 bp')
    args = parser.parse_args()

    main(args)

File: revc/revc.pl
#!/usr/bin/env perl

# PODNAME: revc.pl
# ABSTRACT: Complementing a Strand of DNA

## Author     : ian.sealy@sanger.ac.uk
## Maintainer : ian.sealy@sanger.ac.uk
## Created    : 2014-04-10

use warnings;
use strict;
use autodie;
use Getopt::Long;
use Pod::Usage;
use Carp;
use version; our $VERSION = qv('v0.1.0');

use File::Slurp;

# Default options
my $dataset_file;
my ( $debug, $help, $man );

# Get and check command line options
get_and_check_options();

# Get input
my $s = read_file($dataset_file);
chomp $s;

# Reverse complement
my $sc = reverse $s;
$sc =~ tr/ACGT/TGCA/;

# Print output
printf "%s\n", $sc;

# Get and check command line options
sub get_and_check_options {

    # Get options
    GetOptions(
        'dataset_file=s' => \$dataset_file,
        'debug'          => \$debug,
        'help'           => \$help,
        'man'            => \$man,
    ) or pod2usage(2);

    # Documentation
    if ($help) {
        pod2usage(1);
    }
    elsif ($man) {
        pod2usage( -verbose => 2 );
    }

    # Check options
    if ( !$dataset_file ) {
        pod2usage("--dataset_file must be specified\n");
    }

    return;
}

__END__
=pod

=encoding UTF-8

=head1 NAME

revc.pl

Complementing a Strand of DNA

=head1 VERSION

version 0.1.0

=head1 DESCRIPTION

This script is given "A DNA string s of length at most 1000 bp" and returns "The
reverse complement sc of s".

=head1 EXAMPLES

    perl revc.pl --dataset_file sample.txt

=head1 USAGE

    revc.pl
        [--dataset_file file]
        [--debug]
        [--help]
        [--man]

=head1 OPTIONS

=over 8

=item B<--dataset_file FILE>

File containing "A DNA string s of length at most 1000 bp".

=item B<--debug>

Print debugging information.

=item B<--help>

Print a brief help message and exit.

=item B<--man>

Print this script's manual page and exit.

=back

=head1 DEPENDENCIES

None

=head1 AUTHOR

=over 4

=item *

Ian Sealy <ian.sealy@sanger.ac.uk>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Genome Research Ltd.

This is free software, licensed under:

  The GNU General Public License, Version 3, June 2007

=cut
